pkg/operator/ceph/csi/operator_config.go

```diff
@@ -0,0 +1,162 @@
+limitations under the License.
+*/
+
+package csi
+
+import (
+	csiopv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	opcontroller "github.com/rook/rook/pkg/operator/ceph/controller"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "k8s.io/api/core/v1"
+	k8scsiv1 "k8s.io/api/storage/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const opConfigCRName = "ceph-csi-operator-config"
+
+func (r *ReconcileCSI) CreateOrUpdateOpConfig(csiClusterConfig *cephclient.ClusterInfo, cluster cephv1.CephCluster) error {
+	logger.Info("Creating ceph-CSI operator config CR")
+
+	opConfig := &csiopv1a1.OperatorConfig{}
+	opConfig.Name = opConfigCRName
+	opConfig.Namespace = r.opConfig.OperatorNamespace
+	opConfig.Spec = *r.CreateOpConfigSpec(csiClusterConfig, cluster, opConfig)
+
+	err := r.client.Create(r.opManagerContext, opConfig)
+	if err != nil {
+		if !kerrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "failed to create CSI-operator operator config CR %s", opConfig.Name)
+		}
+		logger.Infof("CSI-operator operator config cr %s already exist, updating...", opConfig.Name)
+		err := r.client.Get(r.opManagerContext, types.NamespacedName{Name: opConfigCRName, Namespace: r.opConfig.OperatorNamespace}, opConfig)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get CSI-operator operator config CR %s", opConfigCRName)
+		}
+
+		opConfig.Spec = *r.CreateOpConfigSpec(csiClusterConfig, cluster, opConfig)
+
+		err = r.client.Update(r.opManagerContext, opConfig)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update CSI-operator operator config CR %s", opConfig.Name)
+		}
+
+		logger.Info("Successfully updated ceph-CSI operator config CR")
+	}
+
+	return nil
+}
+
+func (r *ReconcileCSI) CreateOpConfigSpec(csiClusterConfig *cephclient.ClusterInfo, cluster cephv1.CephCluster, opConfig *csiopv1a1.OperatorConfig) *csiopv1a1.OperatorConfigSpec {
+	maxLogSize, _ := opcontroller.GetLogRotateConfig(cluster.Spec)
+	cephfsClientType := csiopv1a1.KernelCephFsClient
+	if CSIParam.ForceCephFSKernelClient == "false" {
+		cephfsClientType = csiopv1a1.FuseCephFsClient
+	}
+
+	var periodicity csiopv1a1.PeriodicityType
+	switch cluster.Spec.LogCollector.Periodicity {
+	case "1h", "hourly":
+		periodicity = csiopv1a1.HourlyPeriod
+	case "weekly":
+		periodicity = csiopv1a1.WeeklyPeriod
+	case "monthly":
+		periodicity = csiopv1a1.MonthlyPeriod
+	default:
+		periodicity = csiopv1a1.DailyPeriod
+	}
+
+	opConfig.Spec = csiopv1a1.OperatorConfigSpec{
+		LogLevel: int(CSIParam.LogLevel),
+		DriverSpecDefaults: &csiopv1a1.DriverSpec{
+			Log: &csiopv1a1.LogSpec{
+				LogLevel: int(CSIParam.LogLevel),
+				Rotation: &csiopv1a1.LogRotationSpec{
+					MaxLogSize:  maxLogSize,
+					Periodicity: periodicity,
+				},
+			},
+			//  ImageSet: ,
```

Comment:
we need a dicussion on this one, whether Rook will generate the configmap or csi operator should expose the image name names in the CR? cc @travisn

Reply:
Whenever the user needs to specify values, they should be in CRs since that is the expected API for desired state. Configmaps should generally only be used for internal implementation or advanced or uncommon scenarios where we don't really want to expose the settings to users. While the most common is to use the default images, it is common enough to need to specify the images that the CRs would be much preferred IMO. 

That said, the list of images during deployment is a special type of setting since it's a build artifact. And for the build artifacts, the configmap can simplify the downstream deployment to have the images all specified in a configmap for that purpose, with no other user settings. If the images were in the CRs, they would be intermixed with other settings. Ideally I would imagine a fallback mechanism with this priority:
1. Images specified in the configmap
2. Images specified in the CRs
3. Default images

This allows specifying images in the CRs, but also to use the configmap if desired to simplify deployment.

In any case, this is only needed for the stable driver, no need for this feature with the experimental driver.

Reply:
@Madhu-1 ^^

Reply:
@travisn We already had a long conversation, with you participating, where that was discussed and where we justified why we decided to point to an external configmap instead of embedding the images in the CR. 

I am not sure why this issue is reopened again as the decision was made after considering all of the considerations mentioned in the above comment.

Reply:
I know we agreed on the configmap, but I don't recall agreeing why it is not possible to have settings also supported in the CRs with a fallback approach. Please do remind us all. In any case, it's a question for the later stable release, and not needed for experimental mode.

Reply:
There are 3 main reasons:

1. All of our settings including image settings have merge semantics between the different layers (hardcoded, operatorConfig spec, and driver spec). As it is today, it is already complex enough for a user to understand why the operator selected a specific image in a particular container. Introducing images in the CR adds two new layers to that merge/select process - one on the driver spec embedded in the operatorConfig and one in the actual driver spec. 

2. One of the API design principles for ceph-csi-operator is: **"There is only a single way to configure for a desired result"**. Having the ability to configure images on the CR and in a config map has the same target effect and reach. As such we found it a redundant option. 
To make my point more clear, consider the following counter-example: The option to point to an image set from the operator config API vs pointing to an image set from the driver API. In this case, there is a concrete difference, one affects all drivers whereas the other affects a single driver and the user will select which one he would prefer to use based on their desired outcome. 

3. We decided to go with a config map approach vs an inline (in the CR) approach for two main reasons. We desired the ability of orchestration code to be ignorant of the content of the config map - so the config map could be constructed somewhere else (e.g. some dev-op or release CI), and we wanted to separate access control between the user who manages the config map vs the one who configures CSI